.NET 6 config Proxy Server/1

Rick-Anderson opened the pull request:
Fixes #24487 Fixes #18532
[Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-6.0&branch=pr-en-us-24489) -- [Current version of sxs comparison](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer)

MohammadAsadi1372 commented:
https://github.com/MohammadAsadi1372/MohammadAsadi.MasterService

serpent5 commented on aspnetcore/host-and-deploy/proxy-load-balancer.md:
```suggestion
`X-Forwarded-For` is added automatically. For more information, see [Apache Module mod_proxy: Reverse Proxy Request Headers](https://httpd.apache.org/docs/2.4/mod/mod_proxy.html#x-headers). For information on how to forward the `X-Forwarded-Proto` header, see <xref:host-and-deploy/linux-apache#configure-apache>.
```

serpent5 commented on aspnetcore/host-and-deploy/proxy-load-balancer.md:
````diff
@@ -120,19 +113,15 @@ services.Configure<ForwardedHeadersOptions>(options =>
-```
+[!code-csharp[](~/host-and-deploy/proxy-load-balancer/6.1samples/WebPS/Program.cs?name=snippet_https&highlight=14-18)]
 
 This code can be disabled with an environment variable or other configuration setting in a development or staging environment.
````
```suggestion
This code can be disabled with an environment variable or other configuration setting in a development or staging environment:
```

serpent5 commented on aspnetcore/host-and-deploy/proxy-load-balancer.md:
````diff
@@ -172,154 +161,52 @@ app.Use((context, next) =>
-    }
-});
-```
+If the proxy isn't base64-encoding the certificate (as is the case with Nginx), set the `HeaderConverter` option. Consider the following example :
````
```suggestion
If the proxy isn't base64-encoding the certificate, as is the case with Nginx, set the `HeaderConverter` option. Consider the following example:
```

serpent5 commented on aspnetcore/host-and-deploy/proxy-load-balancer/6.1samples/WebPS/Program.cs:
```diff
@@ -0,0 +1,468 @@
+
+var app = builder.Build();
+
+app.Use((context, next) =>
+{
+    context.Request.Scheme = "https";
+    return next();
+});
```
This can use the [new overload](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware/?view=aspnetcore-6.0#prefer-appuse-overload-that-requires-passing-the-context-to-next):

```suggestion
app.Use((context, next) =>
{
    context.Request.Scheme = "https";
    return next(context);
});
```

There's a few others in inline code that I can't get with suggestions that you might want to change to match.

serpent5 commented on aspnetcore/host-and-deploy/proxy-load-balancer/6.1samples/WebPS/Program.cs:
```diff
@@ -0,0 +1,468 @@
+
+if (!app.Environment.IsProduction())
+{
+    app.Use((context, next) =>
+    {
+        context.Request.Scheme = "https";
+        return next();
+    });
```
```suggestion
    app.Use((context, next) =>
    {
        context.Request.Scheme = "https";
        return next(context);
    });
```

serpent5 commented on aspnetcore/host-and-deploy/proxy-load-balancer/6.1samples/WebPS/Program.cs:
```diff
@@ -0,0 +1,468 @@
+
+app.UseForwardedHeaders();
+
+ILogger _logger = app.Logger;
+
+app.Use(async (context, next) =>
+{
+    // Request method, scheme, and path.
+    _logger.LogInformation("Request Method: {Method}", context.Request.Method);
+    _logger.LogInformation("Request Scheme: {Scheme}", context.Request.Scheme);
+    _logger.LogInformation("Request Path: {Path}", context.Request.Path);
+
+    // Headers
+    foreach (var header in context.Request.Headers)
+    {
+        _logger.LogInformation("Header: {Key}: {Value}", header.Key, header.Value);
+    }
+
+    // Connection: RemoteIp
+    _logger.LogInformation("Request RemoteIp: {RemoteIpAddress}",
+        context.Connection.RemoteIpAddress);
+
+    await next();
+});
```
```suggestion
app.Use(async (context, next) =>
{
    // Request method, scheme, and path.
    _logger.LogInformation("Request Method: {Method}", context.Request.Method);
    _logger.LogInformation("Request Scheme: {Scheme}", context.Request.Scheme);
    _logger.LogInformation("Request Path: {Path}", context.Request.Path);

    // Headers
    foreach (var header in context.Request.Headers)
    {
        _logger.LogInformation("Header: {Key}: {Value}", header.Key, header.Value);
    }

    // Connection: RemoteIp
    _logger.LogInformation("Request RemoteIp: {RemoteIpAddress}",
        context.Connection.RemoteIpAddress);

    await next(context);
});
```

serpent5 replied:
This suggestion removes the `_` from `_logger` as it's a local variable. It also passes `context` into the call to `Use`:

```suggestion
var logger = app.Logger;

app.Use(async (context, next) =>
{
    // Request method, scheme, and path.
    logger.LogInformation("Request Method: {Method}", context.Request.Method);
    logger.LogInformation("Request Scheme: {Scheme}", context.Request.Scheme);
    logger.LogInformation("Request Path: {Path}", context.Request.Path);

    // Headers
    foreach (var header in context.Request.Headers)
    {
        logger.LogInformation("Header: {Key}: {Value}", header.Key, header.Value);
    }

    // Connection: RemoteIp
    logger.LogInformation("Request RemoteIp: {RemoteIpAddress}",
        context.Connection.RemoteIpAddress);

    await next(context);
});
```

Rick-Anderson commented:
@Tratcher do you want to take a quick look? Updated your topic for .NET 6 and moved inline snippets to code. In  most case I show the entire `Program.cs` file with relevant code highlighted.  Previous version would just inline snippets like
```
services.Configure<ForwardedHeadersOptions>(options =>
{
    options.ForwardLimit = 2;
    options.KnownProxies.Add(IPAddress.Parse("127.0.10.1"));
    options.ForwardedForHeaderName = "X-Forwarded-For-My-Custom-Header-Name";
});
```
Per @serpent5  review, replaced all `return next();` with `return next(context);`

Tratcher commented on aspnetcore/host-and-deploy/proxy-load-balancer.md:
Linking to the code or the api doc for the middleware isn't much help, neither are for direct consumption. Linking to the ForwardedHeadersOptions API doc is more helpful.

Tratcher commented on aspnetcore/host-and-deploy/proxy-load-balancer.md:
Here too, and for the rest of the doc.

Tratcher commented on aspnetcore/host-and-deploy/proxy-load-balancer/6.1samples/WebPS/Program.cs:
```diff
@@ -0,0 +1,468 @@
+
+app.Run(async (context) =>
+{
+    context.Response.ContentType = "text/plain";
```
We should look at replacing this sample with the new Request Logging Middleware (in a separate issue/PR).

Rick-Anderson replied:
> We should look at replacing this sample with the new Request Logging Middleware (in a separate issue/PR).

Created #24530 to track this. Can you point me to a sample I can use? I assume that's for the next sample too.

Rick-Anderson commented:
@Tratcher  something like the following:

```
using Microsoft.AspNetCore.HttpLogging;
using Microsoft.AspNetCore.HttpOverrides;

var builder = WebApplication.CreateBuilder(args);

builder.Services.AddRazorPages();
builder.Services.AddHttpLogging(options =>
{
    options.LoggingFields = HttpLoggingFields.ResponsePropertiesAndHeaders |
                            HttpLoggingFields.ResponsePropertiesAndHeaders;
});

builder.Services.Configure<ForwardedHeadersOptions>(options =>
{
    options.ForwardedHeaders =
        ForwardedHeaders.XForwardedFor | ForwardedHeaders.XForwardedProto;
});

var app = builder.Build();

app.UseForwardedHeaders();

if (!app.Environment.IsDevelopment())
{
    app.UseExceptionHandler("/Error");
    app.UseHsts();
}

app.UseHttpsRedirection();
app.UseStaticFiles();

app.UseRouting();

app.UseAuthorization();

app.MapRazorPages();

app.Run();
```
